port intraprocedural analysis and fix the long tail..

swissiety opened the pull request:
for deca and other more elaborated BodyInterceptors that were added to soot in the meantime.

- [x] check if the used PriorityQueue is a better implementation than one from a Collections.
- [x] fix changing complete RightOp in ReplaceUseStmtVisitor
- [x] handle nulls e.g. from incomplete information in TypeAssigner
- [x] fix naming of Locals that are not referring to fields (i.e. the one starting with $ - which are now starting that way)
- [x] improve ClassHierarchy - introduce `FallsThroughStmt` and make `Stmt`, `BranchingStmt` and *new* `FallsThroughStmt` Interfaces 
- [x] fix invalid internal state from `MutableStmtGraph.insertBefore(...)`
- [x] fix Aggregator
- [x] fix DeadAssignmentEliminator (it seems to remove more than necessary / does remove a complete branch-target of an IfStmt without fixing the missing branch - resulting in an invalid StmtGraph)
 - [x] ~~fix LocalSplitter - seems it starts already here - (or better port the newer/improved one in Soot?) as it does not (replace all former Local references in Stmts?) anyway it creates a precondition in a way that the DeadAssignmentEliminator wants to remove stmts which it does not when the LocalSplitter is not called. Which means it has leftovers in the StmtGraph or does something else wrong. Have a look at the newest LocalSplitter in Soot before fixing the current port as it seems to be a more precise algorithm.~~ couldnt reproduce this assumption - deadassignmenteliminator failed in the mentioned commit already on its own
- [x] introduces *new parameters to the putEdge() method* for modifying the `Stmtgraph` edges which gives the developer the necessary power to modify the correct successor of a BranchingStmt - putEdge(FallsThroughStmt, Stmt) works like before and putEdge(Stmt, int, Stmt) needs a successorIdx according to the semantics of the respective BranchingStmt. 
- [x] As a sideeffect we can introduce: putEdge(BranchingStmt, int, Stmt) to replace `putEdge(Stmt, int, Stmt)` all Stmts that have no successor like return/throw can't be used as a first Argument as they would fail already at compiletime.
  
closes #667, #692

swissiety committed:
initial commit for intraprocedural analysis

swissiety committed:
progress on priority queue backed by indexed universe

swissiety committed:
refactor copy() as a ctor

swissiety committed:
refactor getDefsOfLocal; JFieldref as leftOp of an assignment is not a Local..

swissiety committed:
improve comment

swissiety committed:
set default BodyInterceptors

swissiety committed:
add comment

swissiety committed:
remove obsolete BodyGenerator reference which could lead to confusion, as it is used with another instance directly in the ClassSources, too

swissiety committed:
style

swissiety commented:
- sets default ByteCodeClassLoadingOptions i.e. BodyInterceptor to create reasonable Jimple (still more Interceptors needed to port from Soot to improve the Jimple generation)
- fixes multiple Issues in Aggregator / Copypropagator / TypeAssigner
- we need to build sootUp more robust i.e. write algorithms that fail early if necessary but work as far as possible if we have incomplete information (we could `log.info()` in those cases were we detect missing information or implement a more elaborated system for imprecisions..?)
- [x] test default bodyinterceptors on ~different~ real input~s~

swissiety committed:
make Aggregator work as in the description

swissiety committed:
replace generics in JIdentityStmt, JAssignStmt with covariant return types and moving fields to subclasses to remove necessity of internal casts

swissiety committed:
fix MutableBlockStmtGraph.insertBefore() : indexing of Stmt->Block was missing

swissiety committed:
fix Interceptors.. and Introduce a new marker interface LhsValuewhich denotes Values that can be used on the left handside of an Assignment - eliminates another runtime check for allowed variable ie leftOp types in JAssignStmt

swissiety committed:
fix LhsValue leftovers

swissiety committed:
fix TypeResolver handling of incomplete information in View

swissiety committed:
fix wrong assumption - and simplified the code.

swissiety commented:
- [x] fix AugEvalTest
- [x] ~one of the Interceptors~ the Aggregator scrambles the code order..
 e.g.
```
           r2 := @this: target.exercise2.FileNotClosedAliasing
	   specialinvoke $stack3.<target.exercise2.File: void <init>()>()
	   r0 = new target.exercise2.File
```
and i hope its only the (already) wrong order that makes a leftover of the $stack3 which should be the r0 as well..

swissiety committed:
more LhsValue typesafety..

swissiety committed:
fix porting bug: soot used getUses() and not getDefs()

swissiety committed:
refactor: LhsValue to LValue; Local.getDefsOfLocal(...) to Local.getDefs(...)

struce2 commented on sootup.core/src/main/java/sootup/core/model/Body.java:
```diff
@@ -587,12 +578,12 @@ public String toString() {
       List<Value> defs = stmt.getDefs();
       for (Value value : defs) {
-        if (value instanceof Local) {
+        if (value instanceof LhsValue) {
```
So, the below code `allDefs.put((Local) value, localDefs);` in line `592`   should be `allDefs.put((LhsValue) value, localDefs);` ?

struce2 commented:
When I use commit `b789f9a153ffb9ba42a64d686be7863a9cdbee2d`, it throw an exception, the error message as follows
```
	... 14 more
Caused by: java.lang.IllegalStateException: if $stack12 != 0: If must have '2' outgoing flow but has '1'.
	at sootup.core.graph.StmtGraph.validateStmtConnectionsInGraph(StmtGraph.java:215)
	... 15 more
Caused by: java.lang.IllegalStateException: if $stack12 != 0: If must have '2' outgoing flow but has '1'.
```

~maybe `ConditionalBranchFolder` BodyInterceptor has an error, but I don't how t fix it :sweat_smile:.~

`MutableStmtGraph.successors(JIfStmt)` returned List size is 0 or 1, so maybe the stmtGraph generater has an error...

swissiety commented:
> When I use commit `b789f9a153ffb9ba42a64d686be7863a9cdbee2d`, it throw an exception, the error message as follows
> 
> ```
> 	... 14 more
> Caused by: java.lang.IllegalStateException: if $stack12 != 0: If must have '2' outgoing flow but has '1'.
> 	at sootup.core.graph.StmtGraph.validateStmtConnectionsInGraph(StmtGraph.java:215)
> 	... 15 more
> Caused by: java.lang.IllegalStateException: if $stack12 != 0: If must have '2' outgoing flow but has '1'.
> ```
> 
> ~maybe `ConditionalBranchFolder` BodyInterceptor has an error, but I don't how t fix it 😅.~
> 
> `MutableStmtGraph.successors(JIfStmt)` returned List size is 0 or 1, so maybe the stmtGraph generater has an error...

please share the complete StackTrace.

swissiety committed:
progress

struce2 commented:
> > When I use commit `b789f9a153ffb9ba42a64d686be7863a9cdbee2d`, it throw an exception, the error message as follows
> > ```
> > 	... 14 more
> > Caused by: java.lang.IllegalStateException: if $stack12 != 0: If must have '2' outgoing flow but has '1'.
> > 	at sootup.core.graph.StmtGraph.validateStmtConnectionsInGraph(StmtGraph.java:215)
> > 	... 15 more
> > Caused by: java.lang.IllegalStateException: if $stack12 != 0: If must have '2' outgoing flow but has '1'.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ~maybe `ConditionalBranchFolder` BodyInterceptor has an error, but I don't how t fix it 😅.~
> > `MutableStmtGraph.successors(JIfStmt)` returned List size is 0 or 1, so maybe the stmtGraph generater has an error...
> 
> please share the complete StackTrace.

When enable `ConditionalBranchFolder` BodyInterceptor, stacktrace as follows

```
java.lang.IllegalStateException: Failed to apply sootup.java.bytecode.interceptors.ConditionalBranchFolder@2b27cc70 to <org.apache.jsp.java_detection_samples: void _jspService(javax.servlet.http.HttpServletRequest,javax.servlet.http.HttpServletResponse)>
	at sootup.java.bytecode.frontend.AsmMethodSource.resolveBody(AsmMethodSource.java:261)
	at sootup.core.model.SootMethod.lazyBodyInitializer(SootMethod.java:98)
	at com.google.common.base.Suppliers$NonSerializableMemoizingSupplier.get(Suppliers.java:181)
	at sootup.core.model.SootMethod.getBody(SootMethod.java:177)
	at main.Main.run(Main.java:62)
	at picocli.CommandLine.executeUserObject(CommandLine.java:1939)
	at picocli.CommandLine.access$1300(CommandLine.java:145)
	at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2358)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2352)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2314)
	at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2179)
	at picocli.CommandLine$RunLast.execute(CommandLine.java:2316)
	at picocli.CommandLine.execute(CommandLine.java:2078)
	at main.Main.main(Main.java:138)
Caused by: java.lang.IndexOutOfBoundsException: Index 1 out of bounds for length 1
	at java.base/jdk.internal.util.Preconditions.outOfBounds(Preconditions.java:64)
	at java.base/jdk.internal.util.Preconditions.outOfBoundsCheckIndex(Preconditions.java:70)
	at java.base/jdk.internal.util.Preconditions.checkIndex(Preconditions.java:266)
	at java.base/java.util.Objects.checkIndex(Objects.java:359)
Caused by: java.lang.IndexOutOfBoundsException: Index 1 out of bounds for length 1

	at java.base/java.util.ArrayList.get(ArrayList.java:427)
	at sootup.java.bytecode.interceptors.ConditionalBranchFolder.interceptBody(ConditionalBranchFolder.java:74)
	at sootup.java.bytecode.frontend.AsmMethodSource.resolveBody(AsmMethodSource.java:258)
	... 13 more

Execution failed for task ':Main.main()'.
> Process 'command '/home/home/applications/java/jdk-17.0.7/bin/java'' finished with non-zero exit value 1

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.

```

when disable  `ConditionalBranchFolder` BodyInterceptor, stacktrace as follows 

```
java.lang.RuntimeException: StmtGraph of <org.apache.jsp.java_detection_samples: void _jspService(javax.servlet.http.HttpServletRequest,javax.servlet.http.HttpServletResponse)> is invalid.
	at sootup.core.model.Body$BodyBuilder.build(Body.java:554)
	at sootup.java.bytecode.frontend.AsmMethodSource.resolveBody(AsmMethodSource.java:264)
	at sootup.core.model.SootMethod.lazyBodyInitializer(SootMethod.java:98)
	at com.google.common.base.Suppliers$NonSerializableMemoizingSupplier.get(Suppliers.java:181)
	at sootup.core.model.SootMethod.getBody(SootMethod.java:177)
	at main.Main.run(Main.java:62)
	at picocli.CommandLine.executeUserObject(CommandLine.java:1939)
	at picocli.CommandLine.access$1300(CommandLine.java:145)
	at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2358)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2352)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2314)
	at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2179)
	at picocli.CommandLine$RunLast.execute(CommandLine.java:2316)
	at picocli.CommandLine.execute(CommandLine.java:2078)
	at main.Main.main(Main.java:138)
Caused by: java.lang.IllegalStateException: visualize invalid StmtGraph: http://magjac.com/graphviz-visual-editor/....
	at sootup.core.graph.StmtGraph.validateStmtConnectionsInGraph(StmtGraph.java:242)
Caused by: java.lang.IllegalStateException: visualize invalid StmtGraph: http://magjac.com/graphviz-visual-editor/...

	at sootup.core.model.Body$BodyBuilder.build(Body.java:552)
	... 14 more
Caused by: java.lang.IllegalStateException: if $stack12 != 0: If must have '2' outgoing flow but has '1'.
	at sootup.core.graph.StmtGraph.validateStmtConnectionsInGraph(StmtGraph.java:215)
	... 15 more
Caused by: java.lang.IllegalStateException: if $stack12 != 0: If must have '2' outgoing flow but has '1'.


Execution failed for task ':Main.main()'.
```
the url is too long, so i deleted something like `?dot=digraph+G+%7B%0A%09compound%3Dtrue%0A%`...

here is classes file

[compiled.zip](https://github.com/soot-oss/SootUp/files/12733476/compiled.zip)

sootUp setup

```
        JavaProject applicationProject =
                JavaProject.builder(new JavaLanguage(8))
                        .addInputLocation(
                                new JavaClassPathAnalysisInputLocation("compiled", SourceType.Application))
                        .build();

        JavaView view = applicationProject.createMutableView();
        view.configBodyInterceptors(analysisInputLocation -> BytecodeClassLoadingOptions.Default);
```

struce2 commented:
disable `ConditionalBranchFolder` BodyInterceptor mean comment out the `new ConditionalBranchFolder()` line in  `BytecodeBodyInterceptors.Default`

swissiety committed:
colon without a mandatory whitespace should be parsed as well

swissiety commented:
Thanks  for the input - it seems the DeadAssignmentEliminator removes one of the flows after the if stmt..
- fix DeadAssignmentEliminator (it seems to remove more than necessary / does remove a complete branch-target of an IfStmt without fixing the missing branch - resulting in an invalid StmtGraph)

seems it starts already in the LocalSplitter as well 
- fix LocalSplitter (or better port the newer/improved one in Soot?) as it does not (replace all former Local references in Stmts?) anyway it creates a precondition in a way that the DeadAssignmentEliminator wants to remove stmts which it does not when the LocalSplitter is not called. Which means it has leftovers in the StmtGraph or does something else wrong. Have a look at the newest LocalSplitter in Soot before fixing the current port as it seems to be a more precise algorithm.

swissiety committed:
remove SourceType.Phantom as it does not make sence in SootUp anymore; remove leftover from port in TypeAssigner

swissiety committed:
add FallsThroughStmt interface; make BranchingStmt and Stmt an interface as well;

swissiety committed:
handle nullType Local naming

swissiety committed:
bugfix input

swissiety committed:
progress; add FallsThroughStmt and refactored Branching/-Stmt

swissiety committed:
refactor methodname

swissiety committed:
fix missing mapping info

swissiety committed:
fix TypeAssigner: don't fail completely if type is not augmentable

swissiety committed:
fix testcases: adapting to fixed local naming..

swissiety committed:
fix tests; add Category

Timbals committed:
fix aggregator combining variables even when conflicting uses exist

swissiety committed:
Merge pull request #700 from Timbals/fix/aggregator

fix aggregator combining variables even when conflicting uses exist

Timbals committed:
improve `removeEdge`

- doesn't always remove all edges when removing at the tail
- returns a boolean indicating if an edge was removed

Timbals committed:
re-enable `modifyStmtToBlockAtTail` test and fix removing edges of loops

Timbals committed:
add a test for block merging in `removeEdge`

Timbals committed:
fix not setting the predecessor correctly when dividing blocks

Timbals committed:
improve `removeNode`

Timbals committed:
test for a conditional followed by a block that gets removed by the `DeadAssignmentEliminator`

Timbals committed:
fix flaky tests by removing usages of `getBlocksSorted`

Timbals committed:
apply formatting

swissiety committed:
Merge pull request #708 from Timbals/fix/MutableBlockStmtGraph

Fix some issues in `MutableBlockStmtGraph`

swissiety committed:
style

swissiety committed:
fix Iterator in getBlocksSorted()

swissiety committed:
wip: fixing Stmt edge connection modification.. theres a problem with non-FallsThroughStmt.. now: giving the modifier more power over branching successors

swissiety committed:
fix core; multiple NPEs

swissiety committed:
adapting tests

swissiety committed:
add deprecation notices..

swissiety commented:
@struce2 can you check if the error still occurs with your input?

struce2 commented:
> @struce2 can you check if the error still occurs with your input?

Works fine without error

swissiety committed:
elaborate api with improved typing: Stmts with zero expected successors can not be added as they are FallsThroughStmt nor BranchingStmt

swissiety committed:
Merge branch 'develop' into add/intraprocedural

# Conflicts:
#	sootup.java.bytecode/src/test/java/sootup/java/bytecode/Soot1577.java
#	sootup.java.bytecode/src/test/java/sootup/java/bytecode/minimaltestsuite/java11/TypeInferenceLambdaTest.java
#	sootup.java.bytecode/src/test/java/sootup/java/bytecode/minimaltestsuite/java8/MethodAcceptingLamExprTest.java

swissiety committed:
fix merge leftover

JonasKlauke commented on sootup.core/src/main/java/sootup/core/jimple/common/stmt/Stmt.java:
```diff
@@ -1,157 +1,58 @@
 package sootup.core.jimple.common.stmt;
 
```
```suggestion
/*-
 * #%L
 * Soot - a J*va Optimization Framework
 * %%
 * Copyright (C) 2018-2023 Linghui Luo, Christian Brüggemann, Markus Schmidt
 * %%
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, either version 2.1 of the
 * License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Lesser Public License for more details.
 * 
 * You should have received a copy of the GNU General Lesser Public
 * License along with this program.  If not, see
 * <http://www.gnu.org/licenses/lgpl-2.1.html>.
 * #L%
 */
```

JonasKlauke commented on sootup.core/src/main/java/sootup/core/model/SourceType.java:
```diff
@@ -23,7 +23,7 @@
-  Application,
-  Library
+  Application, // code that we want to analyze e.g. in call graph generation its traversed
+  Library // code that is given but should not be analyzed e.g. in call graph generation its not
```
This sounds like an issue for the call graph generation. I think we are not ignoring library Sources

JonasKlauke replied:
noted down as issue #715

JonasKlauke commented on sootup.java.bytecode/src/main/java/sootup/java/bytecode/interceptors/BytecodeBodyInterceptors.java:
```diff
@@ -29,21 +29,22 @@
-          new TypeAssigner(),
-    //      new LocalNameStandardizer(),
-          new CopyPropagator(),
+      /*      new CastAndReturnInliner(),
```
This looks strange

JonasKlauke commented on sootup.java.bytecode/src/main/java/sootup/java/bytecode/interceptors/ConditionalBranchFolder.java:
```diff
@@ -83,15 +84,16 @@ public void interceptBody(@Nonnull Body.BodyBuilder builder, @Nonnull View<?> vi
-        builder.removeFlow(predecessor, ifStmt);
-        builder.addFlow(predecessor, tautologicSuccessor);
+        stmtGraph.removeEdge(predecessor, ifStmt);
+        // FIXME: [ms] fix successorIdx
```
```suggestion

```

JonasKlauke commented on sootup.java.bytecode/src/test/java/sootup/java/bytecode/interceptors/AggregatorTest.java:
```diff
@@ -85,16 +140,69 @@ private static Body.BodyBuilder createBodyBuilder(boolean withAggregation) {
+              public List<BodyInterceptor> getBodyInterceptors() {
+                return Arrays.asList(
+                    /*                    new CastAndReturnInliner(),
+                    new UnreachableCodeEliminator(),
```
this also looks strange

JonasKlauke committed:
remove an unnecessary Fix me comment

JonasKlauke committed:
added removed licence header

JonasKlauke committed:
fmt

swissiety committed:
cleanup

swissiety committed:
Merge branch 'develop' into add/intraprocedural

# Conflicts:
#	sootup.java.bytecode/src/test/java/sootup/java/bytecode/interceptors/AggregatorTest.java
